src/utmpx.rs

```diff
@@ -0,0 +1,229 @@
+            target_os = "ios",
+            target_os = "macos"
+        ))]
+        SHUTDOWN_TIME,
+        /// Identifies time after system clock changed.
+        NEW_TIME,
+        /// Identifies time when system clock changed.
+        OLD_TIME,
+        /// Identifies a process spawned by the init process.
+        INIT_PROCESS,
+        /// Identifies the session leader of a logged-in user.
+        LOGIN_PROCESS,
+        /// Identifies a process.
+        USER_PROCESS,
+        /// Identifies a session leader who has exited.
+        DEAD_PROCESS,
+        /// Accounting.
+        #[cfg(target_env = "gnu")]
+        ACCOUNTING,
+    }
+}
+
+impl TryFrom<i16> for EntryType {
+    type Error = Error;
+
+    /// Try to build an `EntryType` from its discriminant.
+    fn try_from(value: i16) -> Result<Self> {
+        match value {
+            libc::EMPTY => Ok(EntryType::EMPTY),
+            #[cfg(target_os = "linux")]
+            libc::RUN_LVL => Ok(EntryType::RUN_LVL),
+            libc::BOOT_TIME => Ok(EntryType::BOOT_TIME),
+            #[cfg(any(
+                target_os = "dragonfly",
+                target_os = "freebsd",
+                target_os = "ios",
+                target_os = "macos"
+            ))]
+            libc::SHUTDOWN_TIME => Ok(EntryType::SHUTDOWN_TIME),
+            libc::OLD_TIME => Ok(EntryType::OLD_TIME),
+            libc::NEW_TIME => Ok(EntryType::NEW_TIME),
+            libc::USER_PROCESS => Ok(EntryType::USER_PROCESS),
+            libc::INIT_PROCESS => Ok(EntryType::INIT_PROCESS),
+            libc::LOGIN_PROCESS => Ok(EntryType::LOGIN_PROCESS),
+            libc::DEAD_PROCESS => Ok(EntryType::DEAD_PROCESS),
+            #[cfg(target_env = "gnu")]
+            libc::ACCOUNTING => Ok(EntryType::ACCOUNTING),
+            _ => Err(Error::invalid_argument()),
+        }
+    }
+}
+
+/// Utmp login record.
+#[derive(Clone, PartialEq, Eq, Debug, Hash)]
+pub struct UtmpEntry {
+    ut_host: String,
+    ut_id: Vec<u8>,
+    ut_line: String,
+    ut_pid: Pid,
+    ut_tv: TimeVal,
+    ut_type: EntryType,
+    ut_user: String,
+}
+
+impl UtmpEntry {
+    /// Return the type of this entry.
+    pub fn entry_type(&self) -> &EntryType {
+        &self.ut_type
+    }
+
+    /// Return the line related to this entry.
+    pub fn line(&self) -> &str {
+        &self.ut_line
+    }
+
+    /// Return the user related to this entry.
+    pub fn user(&self) -> &str {
+        &self.ut_user
+    }
+
+    /// Return the PID related to this entry.
+    pub fn pid(&self) -> &Pid {
+        &self.ut_pid
+    }
+
+    /// Try to build an umtp entry from a raw pointer.
+    fn try_from_ptr(ptr: *mut libc::utmpx) -> Result<Self> {
+        if ptr.is_null() {
+            return Err(Error::invalid_argument());
+        }
+
+        // The lifetime of this whole buffer is very shady and the overall
+        // content small, so we copy the data out and start owning it.
+        let entry = unsafe { *ptr };
```

Comment:
The lifetime can't possibly be shorter than the remainder of this function, or else there will be very big problems.